### DISCUSSION TITLE
请问如何在function或rule里设置目标的平台？ How set platform of target in function/rule?

### DISCUSSION THREAD

**Discussion by traversebitree** [xmake-io/xmake#4702] [Q&A]
由于`Zig c++`对MSVC的支持很有问题，我想写一个rule，判断当某个目标使用`zig`作为工具链时，且目标平台为Windows，不要使用 `xx-windows-msvc`，而是强制使用 `xx-windows-gnu`。其他不采用`zig`作为工具链的目标不受影响，还采用默认的工具链来编译。
> Since `zig c++` support for MSVC is problematic, I want to write a rule that determines that when a target uses `zig` as a toolchain and the target platform is Windows, instead of using `xx-windows-msvc`, it forces the use of `xx-windows-gnu`. Other targets that do not use `zig` as a toolchain are unaffected and are also compiled with the default toolchain.

我只会在function里判断目标的工具链的平台是什么，请问如何在function里设置或修改平台呢？
> I can only determine what the platform of the target's toolchain is in functions, how can I set or modify the platform in functions?

以下为示例程序，在function里使用`target:set_plat("mingw")`命令会报错误"error: field 'set_plat' is not callable (a nil value)"。
> The following is an example programme, using `target:set_plat("mingw")` command in function will report error "error: field 'set_plat' is not callable (a nil value)".

试了`target:set("plat","mingw")`也不行，虽然不报错，但也没起作用
> Tried `target:set("plat", "mingw")` and it doesn't work.

```lua
add_rules("mode.debug", "mode.release")
add_rules("set_mingw_for_zig_on_windows")
add_requires("zig")

rule("set_mingw_for_zig_on_windows")
do
    before_build(function (target)
        if target:toolchains()[1]:config("zig") and target:toolchains()[1]:config("plat")=="windows" then
            target:set_plat("mingw")    -- error: field 'set_plat' is not callable (a nil value)
        end
    end)
end 
rule_end()

target("test_xmake_zig_cxx")
do
    set_kind("binary")
    add_files("src/*.cpp")
    add_toolchains("@zig")
end
target_end()
```

> 参考支持平台链接:  [https://ziglang.org/learn/overview/#:~:text=You%20can%20find%20the%20available%20libc%20targets%20with%20zig%20targets%3A](https://ziglang.org/learn/overview/#:~:text=You%20can%20find%20the%20available%20libc%20targets%20with%20zig%20targets%3A)

**Comment by star-hengxing** [comment]
before_build 换成 on_load/on_config 试试，这些 config 应该在这些阶段才能配置

**Reply by traversebitree** [reply]
谢谢你哈，我最后尝试了把toolchain的plat改了就成功了.
```lua
add_rules("mode.debug", "mode.release")
add_rules("set_mingw_for_zig_on_windows")
add_requires("zig")

rule("set_mingw_for_zig_on_windows")
do
    before_build(function (target)
        if target:toolchains()[1]:config("zig") and target:toolchains()[1]:config("plat")=="windows" then
           target:toolchains()[1]:plat_set("mingw")
        end
    end)
end 
rule_end()

target("test_xmake_zig_cxx")
do
    set_kind("binary")
    add_files("src/*.cpp")
    add_toolchains("@zig")
end
target_end()
```
